### ISSUE TITLE
New charts API

### ISSUE THREAD

**Issue by fpliger** [bokeh/bokeh#2442]
We are formalizing the discussion on the new (more stable) `charts` interface [into this page](https://github.com/bokeh/bokeh/wiki/Bokeh-Days-Working-Document). We can use this issue to discuss the things exposed in the page itself.

**Comment by fpliger** [comment]
> One opportunity with Charts is to build them on top of Blaze. What this would provide is a way to feed a Chart data directly from any abstract data source. This approach is something that will make Bokeh charts uniquely suited to building interactive dashboards, compared to existing capabilities (ggplot, matplotlib, etc.).
>
> Internal => Blaze
> Valid Inputs => Anything Blaze can convert to ColumnDataSource

I'm not sure it's feasible.... I've tried to ping @mrocklin a long time ago, before starting the DataAdapter implementation to check if we could actually use `blaze` expressions. It wasn't the case at the time but things may have changed... 

The big difficulty I see is how do we can leverage on blaze to express the computation we need for some type of charts? If we can solve this and add `blaze.server` to the alchemy then I think we open an amazing scenario of opportunities... We could not only create a `ColumnDataSource` for the charts but also a `BlazeDataSource` that can update if the underlying "data" managed by `blaze.server` changes. 

On a much more concrete scenario I think we should still support the proposed scenarios:

```Python
Chart(df, 'mpg', 'disp', ... )

Chart('./path/to/hdf5::auto', 'mpg', 'disp', ... )

Chart('sqlite://path/to/sqlite::auto', 'mpg', 'disp', ... )
```

but we probably need to use a DataFrame internally and not a blaze expr. @mrocklin @cpcloud what do you think?

@bryevdv @rothnic ^^^

**Comment by mrocklin** [comment]
I suspect that the concern before was that blaze.server doesn't push notifications the way that bokeh does.

The things you've written here though seem quite doable.

**Comment by datnamer** [comment]
This looks really exciting.

Ggplot 2 is cited a fair amount, so It might also be worth looking at Ggvis. The latter's API and data structure improves upon ggplot in several ways:  http://ggvis.rstudio.com/0.1/data-hierarchy.html 

Regarding faceting:  ggvis includes a generalization of faceting called "embeded plots". Hadley's paper on this is: [Visualizing Complex Data with Embedded Plots](http://ggvis.rstudio.com/0.1/data-hierarchy.html)

That sort of paradigm could be a powerful inclusion in bokeh. @rothnic Had some thoughts on this iirc

**Comment by rothnic** [comment]
@fpliger 
One thing that is important to clarify is that there is currently a BlazeDataSource, but it is explicitly for interacting with the blaze server as a specific resource type. What I'm thinking is more abstract, which is focused around removing as many steps as possible for visualizing data for exploratory analysis. Currently, bokeh takes a very direct approach towards getting access to data, which is opposite of the way it deals with graphics. The graphics language has been distilled to models, so we can talk about them on the python, r, or javascript side. I think this should be the same for data, which is especially important for dashboards. So, have the user specify their data resource, whether it is pandas, hdf5, sql, blaze server, etc.

Blaze (and tableau) force you away from bad habits, if you consider that you have no idea of the size or state of the data source. In bokeh, I think we are using the server as an adapter to handle cases which are difficult. Ideally, we'd generate expressions for the computation to be made, then that could be compiled to some form to be executed on the javascript side, but initially at least it would provide a way to support big and arbitrary sources of data. 

I'm working on a concept I plan to prototype soon for a utility for browsing an data source, filtering, and creating a joined data source. I think the output from it could be a data source for one or more Charts. By using blaze we can continue to build on the joined expression for the computations. If you don't approach it this way, then the user would have to build the joined data by hand, then figure out a way to reduce it ahead of time for each specific chart type.

As for feasibility. I have been looking through and I can't see any types of data transforms that we need, that are in pandas and not in blaze. There is definitely lots of functionality in pandas that isn't in blaze, but the charts data transforms are fairly simple. I expect to prototype some more with it soon. The one chart that will be more difficult is Scatter and derivatives, because it isn't aggregating. However, we could look at abstract rendering for that case.

**Comment by rothnic** [comment]
@datnamer, see [faceting](https://github.com/bokeh/bokeh/wiki/Bokeh-Days-Working-Document#facet).

Agree, by definition facets are embeddings. We can embed charts into a grid, or embed charts into pages of a webpage, etc. It is all the same thing. The key functionality a faceting capability is providing is mapping a facet, into an abstract set of coordinates, whether that is (Lon, Lat), (X,Y), or (X, Y, Tab#, Page#).

**Comment by fpliger** [comment]
@rothnic I think we are on the same frequency.. in fact I've played with those guidelines (blaze and pandas) while prototyping  the current `chart` interface data management. The main candidate for the role of `DataAdapter` was actually a pandas DataFrame but at the time we were trying to avoid having such a dependencies for the `charts` interface. This constraint has changed so we can definitely explore this path better.

Another important thing to mention was that blaze was a bit cumbersome for some "simpler" use cases we wanted to support (lists, nested iterables, etc..)... 

Finally a couple of things to consider are that we also want to keep in mind that users may want a hook to create charts with an AjaxDataSource (see #2039) and that now we have definitely a large spectrum of  features that include faceting, composition, aggregation, etc..

That said. I do think it's all feasible.. specially with the more mature versions of bokeh/blaze/blaze.server/myself then 6/7 months ago :wink:

**Comment by birdsarah** [comment]
Thanks for putting this together. Am really excited to see Charts coming together.

I don't have many comments. My main opinion is on 

https://github.com/bokeh/bokeh/wiki/Bokeh-Days-Working-Document#geographical.

I don't particularly like the examples given. This is something that I think ggplot2 doesn't have right. Geography has its own grammar and meaning. 

In general, my gold standard for working with geo data on the web is leafletjs - which I just found a delight to work with. I think we can learn a lot from it.

Am hoping to have a discussion with @brendancol about GIS type stuff soon and think about more low-level objects, which wouldn't necessarily affect charts, but may do.

Do let me know how I can constructively change this mini-ramble into concrete suggestions for the document.

**Comment by rothnic** [comment]
@birdsarah: Thanks, geo is definitely an area that could use attention from some that work more often in it. I recently came from a position where I used it often, but in a 3D use case (see cesiumjs), compared to the more common 2D ones. The thing that might vary somewhat is the web-oriented approach, typically using a smallish dataset of hierarchical js objects, compared to the more typical database-like storage. One thing I'm not experienced with is how more modern gis databases work, how that differs from a traditional row-based relational storage, and how grouping and aggregations work in that case.

In the end, a lot of the functionality detailed for charts comes down to identifying a column/columns to use for grouping, then applying one or more aggregations, applying some visual attribute identifying the group, and/or faceting into multiple plots.

So, where I think we could use help is possibly identifying one or a few typical data structures, then walk through a few examples. This is the type of data structures I've been involved with, but it was more a scientific environment:

**Scientific (xyzt)**
* t, x, y, z, <measures>...
* t, x, y, z, dx, dy, dz, ddx, ddy, ddz, ....<covariance terms>..., <measures>...
* t, lat, lon, altitude, <measures>...

But, you also may have a more traditional business dataset, using a star schema:
* date, user_id, user_name, user_lon, user_lat, clicks, revenue, product_category, product_name

**Comment by fpliger** [comment]
A few comments after reviewing the wiki again:

* https://github.com/bokeh/bokeh/wiki/Bokeh-Days-Working-Document#higher-level-glyphs . One thing to keep in mind is that the same Chart will need to prepare/share data differently (the difference I have in mind now is between Line and Step). That said it's worth exploring the idea...

* https://github.com/bokeh/bokeh/wiki/Bokeh-Days-Working-Document#facet we may want to me more flexible with what's returned from `facet` (for the reasons expressed on the wiki but also because user may want to do N crazy layout arrangements with that...). I think that deciding that we return a grid or a Box is too limiting. We could return a `Facet` object that is just a grid (default) but also easily exposes the single facet plots that users can use to however the want. We could also so some "builtin" presets out  of this object too.. (the proposed `facet_grid`, `facet_wrap`, `facet_tab`, `facet_grid` could be shortcuts to those "presets")

* https://github.com/bokeh/bokeh/wiki/Bokeh-Days-Working-Document#filter-includeexclude is very interesting.. On think we'll need to figure out is how to (or allow) this if the different charts have different `Data` inputs

In general I have some shady thought on some topics that we have there (i.e. Chart Modeling, Column Metatypes,...) but it's better to digest a bit and start working on the implementation. Most of these doubts will probably autoclose at that time..

**Comment by rothnic** [comment]
@fpliger: on higher level glyphs, I think this would fit in for cases like bar chart. A bar is a special case of a rect that is fixed at min of 0, scaled by <agg>(data['field']), and behaves a specific way when stacked/dodged. If you look at the Bar chart builder, there is a lot of detailed manipulation in there to create this behavior that only exists in bar chart. If someone wanted to reuse this behavior, they'd have to reimplement it. In the end, Charts is just a nice interface for automatically generating customized glyphs, based on key identifiers. It should be more focused on identifying the combination of the unique values in the fields, which identify subsets of data, then specifying the unique visual attributes on the specialized glyph. Then it uses these key graphical functions to apply them (stacking, overlaying, dodging, etc.) It seems like how the glyph responds to those operations must be implemented within the glyph, or you need some all knowing thing to do it for you.

On faceting. I agree and should of addressed that. The only thing that faceting is doing is yielding coordinates based on multiple categories. I don't think facet should return a grid at all. It is more abstract in that it isn't tied to how it is displayed. Then there is something that interprets the coordinates into a specific kind of layout, which is where the facet_foo fits in, as you touched on.

**Comment by rothnic** [comment]
Just added references to rough blaze prototyping and some faceting work I did awhile back. Added comments to the blaze prototyping to explain things.

**Comment by rothnic** [comment]
@fpliger: let me know if you are working on data adapter. I've been going through it recently, along with the builder to see what we may need to do for this proposal. If you haven't already completed work on it, I'll take a stab at a prototype this week.

Some discussion areas if anyone wants to provide some feedback:

### Accepted Input Types
One key question is, what about input types that aren't regular table-like structures? Does it make sense to accept both non-grouped and pre-grouped data? In the current implementation, I think accepting both makes things challenging, so I'd lean towards at the beginning only accepting non-grouped data.

So, the three structures we'd accept to start are:

1. Table-like (dataframe, blaze datasource)
2. Lower-level of types representing table-like data
    * Column-like: array, Series, list
    * Row-like: list of dicts (json-like). Can treat this identical to table-like data.

Everything under `2.` would be transformed into a blaze data resource. Note, with column-like data, there is no way to specify a label to group by, so you would have to provide another column-like structure of equal length, or callable that contains the values to be colored by. For example, a logical array would be uniqued to get True/False.

A potential option is that the data-adapter buffers the charts from grouped, or nested data, by transforming it into table-like data (somewhat done currently in data adapter).

### Handling different input types
The builder is just a factory for producing charts. The current approach is to name a function the name of the chart (Scatter), which takes the inputs and initializes the builder (ScatterBuilder).

An alternative approach would be to just call the builder Scatter, then have classmethods defined for handling the off-nominal input types. This would isolate that interface and manipulation into one location. Of course, this would only be needed if we can't easily handle the secondary data types through the normal init method. Just one option I wanted to throw out there.

```python
Scatter.from_dict([{...}, {...}, {...})
Scatter.from_cols([...], [...])
Scatter(data, 'mpg', 'disp')
```

**Comment by datnamer** [comment]
@rothnic , FYI, link to new hadley paper [Visualizing Statistical Models: Removing the Blindfold]( http://onlinelibrary.wiley.com/doi/10.1002/sam.11271/abstract)

**Comment by fpliger** [comment]
@rothnic I was planning to sprint on it this weekend with @chdoig at the EuroPython sprints but we've got too many sprinters coming and I had to follow them (which is good :) 

I'm starting to work on this from today... today I've been collecting info and updating myself (while uploading the EuroPython demos) and I think I'll be able to start playing with a prototype tomorrow. Have you already started something? We definitely should plan the activity on this ..

**Comment by bryevdv** [comment]
This is really good, I think in terms of work flow it might be useful to get some kind of representative screenshot + small code sample reported out daily so that others can very easily follow progress. @fpliger once the two of you get to a point where that is feasible, can you take care of that?

**Comment by fpliger** [comment]
@bryevdv Sure! I was thanking that we can start with the work on the DataAdapter/Input part first as it should still be compatible to the current charts implementation [with minimal efforts], at least with the last branch I was working on. How that sounds?

**Comment by rothnic** [comment]
Some prototyping I've been working on to support the transition to the data and column names interface:
https://gist.github.com/rothnic/a6424b0179bd9b29405b

A large part of the column interface is handling the grouping internal to the chart. This builds on the pandas groupby functionality, providing you the visual attributes for each group as you iterate over them. This provides a generic way to handle this process for one to many chart attributes.

**Comment by rothnic** [comment]
Added a naive implementation of a scatter plot with configurable coloring in the previous gist. Can view the plots here: http://nbviewer.ipython.org/gist/rothnic/a6424b0179bd9b29405b/assign_attrs.ipynb

**Comment by damianavila** [comment]
Really interesting, I will try to take a deeper look later, but I like the spelling a lot! Thanks for sharing it!

**Comment by damianavila** [comment]
Since #2870 was merged, I think this issue server its purpose. Closing it now.
For enhancements or new iterations, please open a new issue/s (if you think it is necessary) and refer this issue at the first post. Thanks!
